src/redismodule.h

```diff
@@ -260,6 +265,142 @@ typedef enum {
+
+typedef struct RedisModuleCommandArg {
+    const char *name;
+    RedisModuleCommandArgType type;
+    int key_spec_index;       /* If type is KEY, this is a zero-based index of
+                               * the key_spec in the command. For other types,
+                               * you may specify -1. */
+    const char *token;        /* If type is PURE_TOKEN, this is the token. */
+    const char *summary;
+    const char *since;
+    int flags;                /* The REDISMODULE_CMD_ARG_* macros. */
+    struct RedisModuleCommandArg *subargs;
+} RedisModuleCommandArg;
+
+typedef struct RedisModuleCommandHistoryEntry {
+    const char *since;
+    const char *changes;
+} RedisModuleCommandHistoryEntry;
+
+typedef struct RedisModuleCommandKeySpec {
+    uint64_t flags; /* REDISMODULE_CMD_KEY_* macros. */
+    RedisModuleKeySpecBeginSearchType begin_search_type;
+    union {
+        struct {
+            /* The index from which we start the search for keys */
+            int pos;
+        } index;
+        struct {
+            /* The keyword that indicates the beginning of key args */
+            const char *keyword;
+            /* An index in argv from which to start searching.
+             * Can be negative, which means start search from the end, in reverse
+             * (Example: -2 means to start in reverse from the panultimate arg) */
+            int startfrom;
+        } keyword;
+    } bs;
+    RedisModuleKeySpecFindKeysType find_keys_type;
+    union {
+        struct {
+            /* Index of the last key relative to the result of the begin search
+             * step. Can be negative, in which case it's not relative. -1
+             * indicating till the last argument, -2 one before the last and so
+             * on. */
+            int lastkey;
+            /* How many args should we skip after finding a key, in order to
+             * find the next one. */
+            int keystep;
+            /* If lastkey is -1, we use limit to stop the search by a factor. 0
+             * and 1 mean no limit. 2 means 1/2 of the remaining args, 3 means
+             * 1/3, and so on. */
+            int limit;
+        } range;
+        struct {
+            /* Index of the argument containing the number of keys to come
+             * relative to the result of the begin search step */
+            int keynumidx;
+            /* Index of the fist key. (Usually it's just after keynumidx, in
+             * which case it should be set to keynumidx + 1.) */
+            int firstkey;
+            /* How many args should we skip after finding a key, in order to
+             * find the next one, relative to the result of the begin search
+             * step. */
+            int keystep;
+        } keynum;
+    } fk;
+} RedisModuleCommandKeySpec;
+
+typedef struct RedisModuleCommandInfo {
+    int version;                  /* REDISMODULE_COMMAND_INFO_VERSION */
```

Comment:
Adding a field to the top-level struct is fine, but adding a field to the arg struct is problematic. Args is an array so the size matters. Key specs too.

Reply:
Ohh, right. Well, making these an array of pointers will defeat the purpose of ease of use. 
I guess we'll have to find another way to provide the extra data when the day comes. 
The other alternative I see is add another spare / reserved pointer in each. 
For key-specs, I'm starting to consider a per-spec description soon.. 
E.G. On SORT we have 3 specs, and it's hard even for us to keep track of what each of them means and why it's there. 
So maybe add 2 reserved pointer in each? (args, and specs)

Reply:
If we want to be able to add fields to these struct arrays, we can instead use pointer arrays. It would allow us to add fields, but AFAIK it kills the possibility to write the whole info in one big nested literal.

Or do you have another solution to this?

Reply:
as i said:
> making these an array of pointers will defeat the purpose of ease of use

so no, i think we should keep it as is, but let's add two spare pointers in each of these structs

Reply:
Discussed this with @oranagra, and another option is for the module to populate the `sizeof` of each array type (possibly hidden behind some macro). On the Redis side, this will allow us to do pointer arithmetic and be able to properly handle the array even if the module's `sizeof` has increased (because it's a newer version).

Reply:
@yossigo any idea how this would look like?

Reply:
Idea: `RedisModule_SetCommandInfo` can be defined as a `static inline` function (or a macro) in `redismodule.h`, which sets the version and size fields and then calls the real (but undocumented) API function.